src/request.rs

```diff
@@ -145,7 +152,13 @@ impl Request {
 
     /// Adds given key and value as query parameter to request url
     /// (resource).
-    pub fn with_param<T: Into<String>, U: Into<String>>(mut self, key: T, value: U) -> Request {
+    ///
+    /// If `urlencoding` is not enabled, it is the responsibility
+    /// of the user to ensure there are no illegal characters in the
+    /// key or value.
+    ///
+    /// If `urlencoding` is enabled, the key and value are both encoded.
+    pub fn with_param<T: AsRef<str>, U: AsRef<str>>(mut self, key: T, value: U) -> Request {
```

Comment:
I've changed the type signature here. While it's not strictly necessary to do so, it saves allocating a new `String` both when `urlencoding` is enable and not. I can remove it if you'd rather not change the API.

I wouldn't anticipate changing from `Into<String>` to `AsRef<str>` to have much of an impact on compatibility off the top of my head

Reply:
There may be other functions that can benefit from this change, but I didn't go looking for them within the scope of this PR